src/metabase/pulse/render/body.clj

```diff
@@ -1030,16 +1030,31 @@
       [:img {:style (style/style {:display :block :width :100%})
              :src   (:image-src image-bundle)}]]}))
 
+(defn- all-unique?
+  [funnel-rows]
+  (let [ks (into #{} (map :key funnel-rows))]
+    (= (count ks) (count funnel-rows))))
+
+(defn- build-funnel-rows
+  [funnel-rows raw-rows]
```

Comment:
There are so many comments in this area, it's hard to comment on the function as a whole. This is a bit of a zoom out comment on the whole function.

I think breaking that readability could be improved by making it clearer that the transformed funnel-rows are only used for their indexing values. The best way to do this I think is to actually discard the other fields and stop calling them funnel-rows.

Also it seems strange to call the result "funnel-rows", as those come in as maps, and the output is a sequence of tuples whose shape is more akin to the `raw-rows`.

```clojure
(defn- reordered-rows [funnel-rows raw-rows]
  (let [idx+keys (map-indexed (fn [idx {k :key, enabled :enabled}]
                                (when enabled [idx k]))
                              funnel-rows)
        foo->bar (into {} (map (fn [row] (vec (take-last 2 row)))) raw-rows)]
    (for [[idx k] idx+keys :when (or idx k)]
      [k (or (get foo->bar k) (get foo->bar idx))])))
```